text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,509 @@
+which is a way of asking the compiler to determine their "intersection"
+(i.e. the period for which both of the loans are active); we then say the
+returned `Item` borrow is valid for that period (which means it may incorporate
+data from both of the input borrows).
+
+## `struct`s and lifetimes
+
+Sometimes you need to build data types that contain borrowed data. Since those
+types can then be used in many contexts, you can't say in advance what the
+lifetime of those borrows will be. Instead, you must take it as a parameter:
+
+```rust
+struct VecIter<'vec, T> {
+    vec: &'vec Vec<T>,
+    index: usize,
+}
+```
+
+Here we're defining a type for iterating over a vector, without requiring
+*ownership* of that vector. To do so, we store a *borrow* of the vector. But
+because our new `VecIter` struct contains borrowed data, it needs to surface
+that fact, and the lifetime connected with it. It does so by taking an explicit
+`'vec` parameter for the relevant lifetime, and using it within.
+
+When using this struct, you can apply explicitly-named lifetimes as usual:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&'vec self) -> VecIter<'vec, T> { ... }
+}
+```
+
+However, in cases like this example, we would normally be able to leave off the
+lifetime with `&`, since there's only one source of data we could be borrowing
+from. We can do something similar with structs:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&self) -> VecIter<'_, T> { ... }
+}
+```
+
+The `'_` marker makes clear to the reader that *borrowing is happening*, which
+might not otherwise be clear.
+
+## `impl` blocks and lifetimes
+
+When writing an `impl` block for a structure that takes a lifetime parameter,
+you can give that parameter a name, which you should strive to make
+*meaningful*:
+
+```rust
+impl<T> VecIter<'vec, T> { ... }
+```
+
+This name can then be referred to in the body:
+
+```rust
+impl<T> VecIter<'vec, T> {
+    fn foo(&self) -> &'vec T { ... }
+    fn bar(&self, arg: &'a Bar) -> &'a Bar { ... }
+}
+```
+
+If the type's lifetime is not relevant, you can leave it off using `'_`:
+
+```rust
+impl<T> VecIter<'_, T> { ... }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+**Note: these changes are designed to *not* require a new epoch**. They do
+expand our naming style lint, however.
+
+## Lifetimes in `impl` headers
+
+When writing an `impl` header, you can mention lifetimes without binding them in
+the generics list. Any lifetimes that are not already in scope (which, today,
+means any lifetime whatsoever) is treated as being bound as a parameter of the
+`impl`.
+
+Thus, where today you would write:
+
+```rust
+impl<'a> Iterator for MyIter<'a> { ... }
+impl<'a, 'b> SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+tomorrow you would write:
+
+```rust
+impl Iterator for MyIter<'iter> { ... }
+impl SomeTrait<'tcx, 'gcx> for SomeType<'tcx, 'gcx> { ... }
```

Comment:
typo: I'm pretty sure you meant
```
impl SomeTrait<'tcx> for SomeType<'tcx, 'gcx> { ... } 
```

Since the above definition had only `SomeTrait<'a>`